react/features/video-quality/subscriber.js

```diff
@@ -289,110 +415,77 @@ function _updateReceiverVideoConstraints({ getState }) {
-                    && !remoteScreenShares.find(id => id === largeVideoParticipantId)) {
-                    quality = getVideoQualityForLargeVideo();
-                }
-                receiverConstraints.constraints[largeVideoSourceName] = { 'maxHeight': quality };
-                receiverConstraints.onStageSources = [ largeVideoSourceName ];
-            }
+    // Tile view.
+    if (shouldDisplayTileView(state)) {
+        if (!visibleRemoteTrackSourceNames?.length) {
+            return;
         }
 
+        visibleRemoteTrackSourceNames.forEach(sourceName => {
+            receiverConstraints.constraints[sourceName] = { 'maxHeight': maxFrameHeightForTileView };
+        });
+
+        // Prioritize screenshare in tile view.
         if (remoteScreenSharesSourceNames?.length) {
-            remoteScreenSharesSourceNames.forEach(sourceName => {
-                receiverConstraints.constraints[sourceName] = { 'maxHeight': VIDEO_QUALITY_LEVELS.ULTRA };
-            });
+            receiverConstraints[sourceNameSignaling ? 'selectedSources' : 'selectedEndpoints']
+                = remoteScreenSharesSourceNames;
         }
 
+    // Stage view.
     } else {
-        receiverConstraints = {
-            constraints: {},
-            defaultConstraints: { 'maxHeight': VIDEO_QUALITY_LEVELS.NONE },
-            lastN,
-            onStageEndpoints: [],
-            selectedEndpoints: []
-        };
-
-        // Tile view.
-        if (shouldDisplayTileView(state)) {
-            if (!visibleRemoteParticipants?.size) {
-                return;
-            }
+        if (!visibleRemoteTrackSourceNames?.length && !largeVideoSourceName && !activeParticipantsSources?.length) {
+            return;
+        }
 
-            visibleRemoteParticipants.forEach(participantId => {
-                receiverConstraints.constraints[participantId] = { 'maxHeight': maxFrameHeight };
+        if (visibleRemoteTrackSourceNames?.length) {
+            visibleRemoteTrackSourceNames.forEach(sourceName => {
+                receiverConstraints.constraints[sourceName] = { 'maxHeight': maxFrameHeightForVerticalFilmstrip };
             });
+        }
 
-            // Prioritize screenshare in tile view.
-            remoteScreenShares?.length && (receiverConstraints.selectedEndpoints = remoteScreenShares);
-
-        // Stage view.
-        } else {
-            if (!visibleRemoteParticipants?.size && !largeVideoParticipantId) {
-                return;
-            }
+        if (getCurrentLayout(state) === LAYOUTS.STAGE_FILMSTRIP_VIEW && activeParticipantsSources.length > 0) {
+            const onStageSources = [ ...activeParticipantsSources ];
 
-            if (visibleRemoteParticipants?.size > 0) {
-                const qualityLevel = getVideoQualityForResizableFilmstripThumbnails(state);
-                const stageParticipantsLevel = getVideoQualityForStageThumbnails(state);
+            activeParticipantsSources.forEach(sourceName => {
```

Comment:
Are `activeParticipantsSources` also included in the `visibleRemoteTrackSourceNames` ? If so, aren't we sending duplicate constraints here ?

Reply:
I think not always. IIRC some of they may be there and some not. But anyway even if there is overlap we will just override the value and not duplicate it which is exactly what we want because the activeParticipantSources are the once displayed on stage.